kopf/reactor/daemons.py

```diff
@@ -0,0 +1,394 @@
+                stopper.set(reason=primitives.DaemonStoppingReason.DAEMON_SIGNALLED)
+                logger.debug(f"Daemon {daemon_id!r} is signalled to exit gracefully.")
+            delays.append(backoff - age)
+
+        elif timeout is not None and age < timeout + (backoff or 0):
+            if not stopper.is_set(reason=primitives.DaemonStoppingReason.DAEMON_CANCELLED):
+                stopper.set(reason=primitives.DaemonStoppingReason.DAEMON_CANCELLED)
+                logger.debug(f"Daemon {daemon_id!r} is signalled to exit by force.")
+                daemon.task.cancel()
+            delays.append(timeout + (backoff or 0) - age)
+
+        elif timeout is not None:
+            if not stopper.is_set(reason=primitives.DaemonStoppingReason.DAEMON_ABANDONED):
+                stopper.set(reason=primitives.DaemonStoppingReason.DAEMON_ABANDONED)
+                logger.warning(f"Daemon {daemon_id!r} did not exit in time. Leaving it orphaned.")
+                warnings.warn(f"Daemon {daemon_id!r} did not exit in time.", ResourceWarning)
+
+        else:
+            logger.debug(f"Daemon {daemon_id!r} is still exiting. Next check is in {polling}s.")
+            delays.append(polling)
+
+    return delays
+
+
+async def daemon_killer(
+        *,
+        memories: containers.ResourceMemories,
+) -> None:
+    """
+    An operator's root task to kill the daemons on the operator's shutdown.
+    """
+
+    # Sleep forever, or until cancelled, which happens when the operator begins its shutdown.
+    try:
+        await asyncio.Event().wait()
+    except asyncio.CancelledError:
+        pass
+
+    # Terminate all running daemons when the operator exits (and this task is cancelled).
+    coros = [
+        stop_daemon(daemon_id=daemon_id, daemon=daemon)
+        for memory in memories.iter_all_memories()
+        for daemon_id, daemon in memory.daemons.items()
+    ]
+    if coros:
+        await asyncio.wait(coros)
+
+
+async def stop_daemon(
+        *,
+        daemon_id: containers.DaemonId,
+        daemon: containers.Daemon,
+) -> None:
+    """
+    Stop a single daemon.
+
+    The purpose is the same as in `stop_resource_daemons`, but this function
+    is called on operator exiting, so there is no multi-step handling,
+    everything happens in memory and linearly (while respecting the timing).
+    """
+    if isinstance(daemon.handler, handlers_.ResourceDaemonHandler):
+        backoff = daemon.handler.cancellation_backoff
+        timeout = daemon.handler.cancellation_timeout
+    elif isinstance(daemon.handler, handlers_.ResourceTimerHandler):
+        backoff = None
+        timeout = None
+    else:
+        raise RuntimeError(f"Unsupported daemon handler: {daemon.handler!r}")
+
+    # Whatever happens with other flags & logs & timings, this flag must be surely set.
+    daemon.stopper.set(reason=primitives.DaemonStoppingReason.OPERATOR_EXITING)
```

Comment:
maybe something like...
```
        if not daemon.task.done():
            daemon.stopper.set(reason=primitives.DaemonStoppingReason.OPERATOR_EXITING)
            daemon.logger.debug(f"Daemon {daemon_id!r} is signalled to exit gracefully.")
```
? Otherwise there will be no log message if the daemon exits instantly.

Reply:
Right. I will take a look at this.